urbit/urbit
# Add Remote Scry Protocol

Resolves https://github.com/urbit/urbit/issues/5788.

TODOs:
- [x] **find out why the compiler is super slow on this branch** EDIT: doesn't seem much slower than other branches
- [x] make corresponding Vere PR
- [x] link to Vere PR here so CI can pass
  - [x] fix CI errors
    - [x] aqua—untested
    - [x] /tests/sys/grq—etch-shot changed the definition of a packet (or .shot) by adding two new flags (req and sam) 
    - [x] /tests/sys/zuse/balk
- [x] fix Clay compilation errors
- [x] stop instantiating bunted `+pe` core in Ames in the `%hunk` scry handler
- [x] check that -keen works correctly 
  - [ ] and its arguments make sense
  - [x] add a way for %spider to track/cancel scry requests. if the thread is cancelled, tell ames to remove this request
    - currently spider tracks any threads that produce a %keen task to %ames
    - If a thread is done, stops or fails, it will remove its ID from the map of threads it keeps in its state, sending clean up cards (not when is done) to ames to remove the scry request
- [x] include outgoing  scry requests to the ~debug dashboard
- [x] add a command to cancel scry request (|pass might work? maybe not—due to how ames handles ducts from vanes, it might not work, so adding a force flag to %yawn that cancels all the scry requests that any vane has requested)
  - `|wham ~ship path` removes all listeners associated with that scry path and gives them `%miss` gifts
- [x] check that scrying into %gall works—will need to merge https://github.com/urbit/urbit/pull/6388
  - [x] remind @belisarius222 to include his testing app
- [x] write -pine thread to test request at latest
  - %pine currently goes through %ames (see +on-plea) so it costs a disk write in the publisher. 
  - it should work for %clay 
  - [x] check for a requester that knows about %pine, and an old publisher
    - The publisher will crash, either because it expects a %cork plea, or it doesn't support %$ as the recipient vane (see https://github.com/urbit/urbit/pull/6406/commits/8387a28bd2075cb38b54b85dca0f19c3e5d8526c) and the requester sinks the naxplanation as expected
  - [x] check requesting a case for a path that has no data
    - The current behavior is, if when we scry for the case we get `[~ ~]`, we nack, the requester will then emit a %miss to the original initiator of the remote scry 
- for gall (see also https://github.com/urbit/urbit/pull/6388)
  - (if you bind something to %gall, that will be public data; otherwise we need a real permission system for agents). 
  - [x] support %w scry (probably not?) (same as for clay that gives you all the cases: %ud.. etc). use case as %gall, what the latest case that's been bound to this scry path
  - [x] %a to scry into of gall
  - [x] ~%y~ %t scry, to get what paths are bound
  - [x] check what happens if you nuke an app? (it should know how to resume if you revive the app)
- [x] add %miss for %keen (e.g. for malformed scry responses)
- [x] check that getting a scry request for ships that cant' handle them is discarded gracefully
  - probably adding something to %ames for handling it. 
  - add teaser support for remote scry for kelvins prior to the full release, so it checks the packet header and no-ops (it looks like vere is already doing this)
  - check the existing behavior, vere might already drop those packets
- [ ] general testing
  - [ ] packetization unit test (low risk)
  - [ ] ames unit test: 
    - idea is to check moves that come of a %keen task, use then to construct the call to the server ames; see tests/sys/grq.hoon)
    - [x]  1-packet remote-scry request/response
      - ~~currently this is failing for some reason. I get a "can't recenter the guard page" error but it doesn't seem to be the test crashing because its  (as far as I can see) emitting the expected moves~~ CI passes again
    - [ ] client that makes an outgoing scry request where the response is a 3 packet message, with a page size of 2 (will need to construct manually)
  - [x] unit test: have two different ducts ask %ames the same request scry path and check that it's not duplicated, and that the response is sent to both.
  - [ ] diagnostic tools
    - [x] added /fine/ducts/[path]
  - [x] ames verbosity (verb/sift)—`|ames-verb %fin`
  - [ ] vere verbosity 
- [x] include the size of the scry cache in |mass
- [x] review the code for general cleanliness (record meeting)
  - [x] ames.hoon 
    - [x] (`%keen`) 
    - [x] (`%miss`)
  - [x] ames.c
  - [x] gall.hoon
  - [x] clay.hoon?
- [X] test on fakezods
- [x] test on moons
- [x] stretch goal: make new `[%$ *]` namespace in Gall consistent across all cares (this includes changing arvo)
- [x] requests are signed, but those signatures are never verified
- [x] life and rift in the fine-specific namespace paths are never checked
- [x] (fixed) :spider is special-cased for %keen cancellation (s/b %give %miss)
- [x] error handling is partial and suspect
- [x] namespace paths are parsed/validated in ad-hoc and confusing ways (see +de-omen call in %clay)
- [x] %fine interface is hard to use
- [x] remove dead code path in clay (%warp-index induced read)
- [x] clean up the names of the remote scry interface
- [x] investigate the crash molting from %larva to %adult ames, booting with -x
- [x] Fix Clay nest-fail on `+load` from version `%12` to this PR
- [x] rename `*-etch-keen` in `ames.hoon` to `-etch-wail`
- [ ] figure out what (if anything) to do about invalid `$path`s in Clay, such as README.md, since they can't be fetched using Fine
- [x] remove signature field from scry requests, replacing with a version byte
- [x] check if `.siz` can be safely removed from `$meow` without causing problems with leading zeros
